### PR TITLE
Add plugin API method UploadFileFromReader

### PR DESCRIPTION
**Pull request by mickmister**
#### Summary

At the moment, the UploadFile plugin API method requires a `[]byte` to be used to upload a file. This has negative implications on servers with a small amount of RAM.

This PR implements an additional plugin API method `UploadFileFromReader` to support uploading via an `io.Reader`.

#### Ticket Link

Fixes https://mattermost.atlassian.net/browse/MM-45060

#### Release Note
```release-note
Plugin API method `UploadFileFromReader` is added to allow plugins to upload files without storing the entire file contents in memory.
```

### CODE REVIEW

File: app/plugin_api.go [hunk 1]
```diff
@@ -759,6 +760,13 @@ func (api *PluginAPI) UploadFile(data []byte, channelID string, filename string)
 	return api.app.UploadFile(api.ctx, data, channelID, filename)
 }
 
+func (api *PluginAPI) UploadFileFromReader(reader io.Reader, channelID string, filename string) (*model.FileInfo, *model.AppError) {
+	return api.app.UploadFileX(api.ctx, channelID, filename, reader,
+		UploadFileSetUserId("nouser"),
```
Comment (agnivade) [inline app/plugin_api.go:765]: Passing "nouser" seems incorrect. I'd suggest to perhaps make `userID` a param itself. Alternatively, there's also `api.ctx.Session().UserId` but that seems to be initialized at startup, so likely will be empty.

Comment (mickmister) [reply app/plugin_api.go:765]: @agnivade My intention was to mimic the implementation of the [UploadFile](https://developers.mattermost.com/integrate/plugins/server/reference/#API.UploadFile) plugin API method, which does not require a user ID. Most methods of the plugin API do not require a user ID, unless the operation absolutely requires the user ID.

I'm open to the idea of requiring the user ID here, though I think this sort of changes the general patterns of the plugin API methods, and we should rethink this with how we want to continue adding new methods. cc @hanzei

Comment (agnivade) [reply app/plugin_api.go:765]: ok

**Review by agnivade — APPROVED** [review]

**Review by hanzei — COMMENTED** [review]
Would you mind adding an E2E test in https://github.com/mattermost/mattermost-server/tree/master/app%2Fplugin_api_tests?

**Comment by mickmister** [comment]
@wiggin77 I suppose it is indeed already implemented via upload sessions. Thanks for mentioning this :+1:

**Pull request closed without merging**